### ISSUE TITLE
Сontinuous backup (PITR) to disk (s3/hdfs/nas/nfs) (Time Travel)

### ISSUE THREAD

**Issue by UnamedRus** [ClickHouse/ClickHouse#30407]
**Use case**
Point-in-time restore

Time Travel
```
SELECT * FROM my_table AT(TIMESTAMP => 'Mon, 01 May 2015 16:20:00 -0700'::timestamp_tz);
SELECT * FROM my_table AT(OFFSET => -60*5);
```

Give ability to run queries against state of table at specific moment of time in past.
With ability to revert back to this table state. (for example revert change of column type, drop column statement or DELETE WHERE)
So it has limited usage for append only tables, but with support for light-weight deletes (or updates), people will modify data more and ability to look in the past will became more useful.

It's also possible to offload outdated parts / columns to cold and cheap storage like s3. (PITR for backup?)

**Describe alternatives you've considered**

filesystem with snapshotting (ZFS and others)
some external tool, which will watch for new parts and upload them to storage
FREEZE TABLE and (or) BACKUP TABLE.

**Additional context**

Snowflake: https://docs.snowflake.com/en/user-guide/data-time-travel.html

Oracle: https://docs.oracle.com/cd/E11882_01/appdev.112/e41502/adfns_flashback.htm#ADFNS1008

CockroachDB: https://www.cockroachlabs.com/blog/time-travel-queries-select-witty_subtitle-the_future/

**Comment by alexey-milovidov** [comment]
Write amplification is about 5x. Therefore, it shouldn't be really continuous.

**Comment by Slach** [comment]
@alexey-milovidov sorry, could you explain please, how write amplification related to point-in-time-restore?

**Comment by alexey-milovidov** [comment]
For point-in-time restore, you have to keep old parts, and the amount of all parts, including old ones, is about 5x.

**Comment by UnamedRus** [comment]
> Write amplification is about 5x
> , and the amount of all parts, including old ones, is about 5x.

1. It's not write amplification, but it's more like "keep data amplification".
2. Snowflake and other vendors allow to travel back in time for limited window, one day for example.
So, it does mean, that we will keep X5 data only for latest day(partition to be precise). (if we keep data for year, it will be not even noticeable) 


* We can keep not old parts by themself, but keep "all" versions of rows 
And here we have 2 options:

#### Normal MergeTree tables, with simple addictive Merge
Basically we just need to have additional column: _creation_csn and that's it, it actually will work even today.

#### ReplacingMergeTree, AggregatingMergeTree.

We will have 2 set of parts: 
* active, which will have result of ReplacingMerge, AggregatingMerge
* inactive, which will have result of simple Merge + _creation_csn column.

**Comment by alex-zaitsev** [comment]
Side note -- Vertica has a notion of epoch, ability to set an epoch explicitly or using some time conditions, and traverse back when needed.

**Comment by UnamedRus** [comment]
Databend (ClickHouse inspired db) also have timetravel 
https://databend.rs/blog/time-travel
